lib/tasks/migrate.rake

```diff
@@ -1,76 +1,145 @@
+PUPPET_LOG_LEVELS = %w[debug info notice warning err alert emerg crit].freeze
+$reports_migrate_running = true
+
 namespace :reports do
-  def keywords(status)
+  def detect_puppet_keywords(status, logs)
     keywords_set = {}
-    keywords_set["PuppetFailed"] = true if status["failed"] == 1
-    keywords_set["PuppetFailedToRestart"] = true if status["failed_restarts"] == 1
-    keywords_set["PuppetCorrectiveChange"] = true if status["corrective_change"] == 1
-    keywords_set["PuppetSkipped"] = true if status["skipped"] == 1
-    keywords_set["PuppetRestarted"] = true if status["restarted"] == 1
-    keywords_set["PuppetScheduled"] = true if status["scheduled"] == 1
-    keywords_set["PuppetOutOfSync"] = true if status["out_of_sync"] == 1
+    # from statuses
+    keywords_set["PuppetFailed"] = true if status["failed"].positive?
+    keywords_set["PuppetFailedToRestart"] = true if status["failed_restarts"].positive?
+    keywords_set["PuppetCorrectiveChange"] = true if status["corrective_change"].positive?
+    keywords_set["PuppetSkipped"] = true if status["skipped"].positive?
+    keywords_set["PuppetRestarted"] = true if status["restarted"].positive?
+    keywords_set["PuppetScheduled"] = true if status["scheduled"].positive?
+    keywords_set["PuppetOutOfSync"] = true if status["out_of_sync"].positive?
+    # from logs
+    logs.each do |level, resource, message|
+      keywords_set["PuppetFailed:#{resource}"] = true if level == "err" && resource != "Puppet"
+    end
+    keywords_set.keys
```

Comment:
We want to stay out of duplicates, granted that uniq might be even better performing because typically we are dealing with just dozens of keywords. Let's do that.